Dockerfile

```diff
@@ -1,34 +1,38 @@
-FROM ipfs/go-ipfs:release
+FROM golang:1.9-stretch AS builder
 MAINTAINER Hector Sanjuan <hector@protocol.ai>
 
-EXPOSE 9094
-EXPOSE 9095
-EXPOSE 9096
+# This build state just builds the cluster binaries
+
+ENV GOPATH     /go
+ENV SRC_PATH   $GOPATH/src/github.com/ipfs/ipfs-cluster
+
+COPY . $SRC_PATH
+WORKDIR $SRC_PATH
+RUN make install
+
+#------------------------------------------------------
+FROM ipfs/go-ipfs
+MAINTAINER Hector Sanjuan <hector@protocol.ai>
+
+# This is the container which just puts the previously
+# built binaries on the go-ipfs-container.
 
 ENV GOPATH     /go
-ENV PATH       /go/bin:$PATH
 ENV SRC_PATH   /go/src/github.com/ipfs/ipfs-cluster
 ENV IPFS_CLUSTER_PATH /data/ipfs-cluster
 
-VOLUME $IPFS_CLUSTER_PATH
+EXPOSE 9094
+EXPOSE 9095
+EXPOSE 9096
 
-COPY . $SRC_PATH
+COPY --from=builder $GOPATH/bin/ipfs-cluster-service /usr/local/bin/ipfs-cluster-service
+COPY --from=builder $GOPATH/bin/ipfs-cluster-service /usr/local/bin/ipfs-cluster-ctl
+COPY --from=builder $SRC_PATH/docker/entrypoint.sh /usr/local/bin/start-daemons.sh
 
-RUN apk add --no-cache --virtual cluster-deps make musl-dev go git \
-    && apk add --no-cache jq \
-    && go get -u github.com/whyrusleeping/gx \
-    && go get -u github.com/whyrusleeping/gx-go \
-    && cd $SRC_PATH \
-    && gx install --global \
-    && gx-go rewrite \
-    && go build \
-    && make -C ipfs-cluster-service install \
-    && make -C ipfs-cluster-ctl install \
-    && cp docker/entrypoint.sh /usr/local/bin/start-daemons.sh \
-    && chmod +x /usr/local/bin/start-daemons.sh \
-    && apk del --purge cluster-deps \
-    && cd / && rm -rf /go/src /go/bin/gx /go/bin/gx-go
+RUN mkdir -p $IPFS_CLUSTER_PATH && \
+    chown 1000:100 $IPFS_CLUSTER_PATH
 
+VOLUME $IPFS_CLUSTER_PATH
 ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/start-daemons.sh"]
 
 CMD ["$IPFS_CLUSTER_OPTS"]
```

Comment:
Out of curiousity, where does $IPFS_CLUSTER_OPTS get set?

Reply:
You could set them when running docker `docker run --env IPFS_CLUSTER_OPTS="-d"...`. But if you do `docker run ipfs-cluster <arguments>` the arguments will replace the CMD anyway. I think it just allows some more flexibility.